File: WebView2/cs/JavaScriptMusicSample/NativeMediaPlayer/PlaylistDataFetcher.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using System;
using System.Net;
using System.Threading.Tasks;
using Windows.Storage;

namespace NativeMediaPlayer
{
    /// <summary>
    /// This class is responsible for retrieving playlist JSON data for use by the MediaPlaybackController.
    /// 
    /// In this sample, playlists are stored in static json files inside the app package, but they could
    /// easily be pointed at web URLs which allow a server to construct them dynamically instead.
    /// </summary>
    static class PlaylistDataFetcher
    {
        /// <summary>
        /// Returns JSON data describing all tracks in a particular playlist. See the playlistdata folder
        /// in the main project for an example.
        /// For this sample, the Id is simply its filename.
        /// </summary>
        /// <param name="playlistId">A unique identifier for the playlist to describe.</param>
        public static async Task<string> GetPlaylistTracks(string playlistId)
        {
            Uri uri = new Uri($"ms-appx:///WebCode/playlistdata/{playlistId}.json");
            return await FetchStringFromUri(uri);
        }

        /// <summary>
        /// Constructs a URI to a particular track, given its Id.
        /// For this sample, the Id is simply its filename.
        /// </summary>
        /// <param name="trackId">A unique identifier for the track.</param>
        /// <returns>A fully-qualified URI that can be used to access the track.</returns>
        public static string GetUriFromTrackId(string trackId)
        {
            return $"ms-appx:///WebCode/music/{trackId}.mp3";
        }

        /// <summary>
        /// Helper function to retrieve the contents of a text file either from local storage or the web.
        /// </summary>
        /// <param name="uri">URI to the file to retrieve.</param>
        /// <returns>A string containing the contents for the file.</returns>
        private static async Task<string> FetchStringFromUri(Uri uri)
        {
            if (uri.Scheme.Equals("ms-appx"))
            {
                StorageFile file = await StorageFile.GetFileFromApplicationUriAsync(uri);
                return await FileIO.ReadTextAsync(file);
            }
            else
            {
                using (WebClient wc = new WebClient())
                {
                    return await wc.DownloadStringTaskAsync(uri);
                }
            }
        }
    }
}

File: WebView2/cpp/JavaScriptMusicSample/NativeMediaPlayer/PlaylistDataFetcher.cpp
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

#include "pch.h"
#include "PlaylistDataFetcher.h"
#include "PlaylistDataFetcher.g.cpp"
#include <sstream>
#include <winrt/Windows.Storage.h>

using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::Storage;
using namespace winrt::Windows::Web::Http;

namespace winrt::NativeMediaPlayer::implementation
{
    winrt::Windows::Foundation::IAsyncOperation<hstring> PlaylistDataFetcher::GetPlaylistTracks(hstring playlistId)
    {
        std::wostringstream strStream{};
        strStream << L"ms-appx:///WebCode/playlistdata/" << playlistId.c_str() << L".json";
        Uri uri{ strStream.str() };
        hstring str{ co_await FetchStringFromUri(uri) };
        co_return str;
    }
    hstring PlaylistDataFetcher::GetUriFromTrackId(hstring const& trackId)
    {
        std::wostringstream strStream{};
        strStream << L"ms-appx:///WebCode/music/" << trackId.c_str() << L".mp3";
        return strStream.str().c_str();
    }

    /// <summary>
    /// Helper function to retrieve the contents of a text file either from local storage or the web.
    /// </summary>
    /// <param name="uri">URI to the file to retrieve.</param>
    /// <returns>A string containing the contents for the file.</returns>
    IAsyncOperation<hstring> PlaylistDataFetcher::FetchStringFromUri(Uri uri)
    {
        if (uri.SchemeName() == L"ms-appx")
        {
            StorageFile file{ co_await StorageFile::GetFileFromApplicationUriAsync(uri) };
            hstring text{ co_await FileIO::ReadTextAsync(file) };
            co_return text;
        }
        else
        {
            // It is best to avoid re-creating an HttpClient every time a new request is made, for performance.
            // Depending on your scenario, you may need to cusomize the HttpClient's settings further.
            // For more details, see:
            // https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/http/httpclient-guidelines
            static const HttpClient httpClient{};

            HttpRequestMessage getRequest{ HttpMethod::Get(), uri };
            HttpResponseMessage response{ co_await httpClient.SendRequestAsync(getRequest) };
            if (response.IsSuccessStatusCode())
            {
                hstring text{ co_await response.Content().ReadAsStringAsync() };
                co_return text;
            }
            else
            {
                hstring errStr = L"Failed to retrieve string from URI: " + uri.ToString() + L" Failure Code: " + to_hstring(static_cast<int>(response.StatusCode()));
                OutputDebugString(errStr.c_str());
                co_return L"";
            }
        }
    }
}
